controller/system_backup_controller.go

```diff
@@ -0,0 +1,1068 @@
+
+			_, err = backupTargetClient.SystemBackupUpload(systemBackup.Name, archievePath, systemBackup.Status.Version, systemBackup.Status.GitCommit, c.managerImage, defaultEngineImage)
+			if err != nil {
+				if types.ErrorAlreadyExists(err) {
+					err = nil
+					continue
+				}
+
+				log.WithError(err).Debugf("uploading %v to backup target", archievePath)
+				continue
+			}
+		}
+	}
+}
+
+func (c *SystemBackupController) cleanupRemoteSystemBackupFiles(systemBackup *longhorn.SystemBackup, backupTargetClient engineapi.BackupTargetClientSystemBackupInterface) (err error) {
+	backupTargetSetting, err := c.ds.GetSetting(types.SettingNameBackupTarget)
+	if err != nil {
+		return err
+	}
+
+	if backupTargetSetting.Value == "" {
+		return nil
+	}
+
+	backupTargetURL := backupTargetClient.GetURL()
+	_, err = backupTargetClient.SystemBackupDelete(systemBackup)
+	if err != nil {
+		return errors.Wrapf(err, "failed to delete %v system backup in %v", systemBackup.Name, backupTargetURL)
+	}
+
+	log := getLoggerForSystemBackup(c.logger, systemBackup.Name, backupTargetURL)
+	log.Infof("Deleted system backup in backup target")
+	return nil
+}
+
+type SystemBackupMeta struct {
+	LonghornVersion       string      `json:"longhornVersion"`
+	LonghornGitCommit     string      `json:"longhornGitCommit"`
+	KubernetesVersion     string      `json:"kubernetesVersion"`
+	LonghornNamespaceUUID string      `json:"longhornNamspaceUUID"`
+	SystemBackupCreatedAt metav1.Time `json:"systemBackupCreatedAt"`
+	ManagerImage          string      `json:"managerImage"`
+}
+
+func (c *SystemBackupController) GenerateSystemBackup(systemBackup *longhorn.SystemBackup, archievePath, tempDir string) {
+	log := getLoggerForSystemBackup(c.logger, systemBackup.Name, "")
+
+	var err error
+	var errMessage string
+	existingSystemBackup := systemBackup.DeepCopy()
+	defer func() {
+		if err != nil {
+			c.cacheStateUpdate(
+				systemBackupRecordTypeError, longhorn.SystemBackupStateError,
+				longhorn.SystemBackupConditionReasonGenerate, errMessage,
+			)
+		} else {
+			c.cacheStateUpdate(
+				systemBackupRecordTypeNormal, longhorn.SystemBackupStateUpload,
+				longhorn.SystemBackupConditionReasonGenerate, fmt.Sprintf(SystemBackupMsgCreatedArchieveFmt, archievePath),
+			)
+		}
+
+		c.handleState(systemBackup, existingSystemBackup, err, log)
+	}()
+
+	namespace, err := c.ds.GetNamespace(systemBackup.Namespace)
+	if err != nil {
+		errMessage = fmt.Sprintf(SystemBackupErrGetFmt, "Longhorn namespace")
+		return
+	}
+
+	kubeVersion, err := c.ds.GetKubernetesVersion()
+	if err != nil {
+		errMessage = fmt.Sprintf(SystemBackupErrGetFmt, "Kubernetes version")
+		return
+	}
+
+	SystemBackupMeta := &SystemBackupMeta{
+		LonghornVersion:       meta.Version,
+		LonghornGitCommit:     meta.GitCommit,
+		KubernetesVersion:     kubeVersion.GitVersion,
+		LonghornNamespaceUUID: string(namespace.UID),
+		SystemBackupCreatedAt: systemBackup.Status.CreatedAt,
+		ManagerImage:          c.managerImage,
+	}
```

Comment:
NIT: you can abstract this as a separate function then move it into `SystemBackupMeta`

Reply:
Addressed.